zip-0311.rst

```diff
@@ -2,7 +2,262 @@
+
+and does not reveal:
+
+- Edward's payment address.
+- Any of Edward's other diversified payment addresses.
+- Any of Edward's other transactions.
+
+Shielded withdrawals from an exchange
+-------------------------------------
+
+CorpBux is an exchange with a transparent hot wallet, and enables customers to withdraw
+their funds to Sapling addresses. They can create a payment disclosure, that they can give
+to their customers, that reveals:
+
+- The correct amount was sent to the customer's recipient address in the given transaction.
+- CorpBux was the sender of the given transaction.
+
+
+Requirements
+============
+
+- The payment disclosure may disclose the contents of zero or more Sapling outputs within
+  a single transaction. (Zero outputs is useful for t2z transactions.)
+- Payment disclosures are non-malleable, and can only be created by a sender of the
+  transaction.
+- A payment disclosure may be tied to a challenge, creating an interactive proof.
+- Senders with a Sapling address may choose to provably reveal that address as part of the
+  payment disclosure.
+- Senders are not required to remember any ephemeral data from the creation of a transaction
+  in order to produce a payment disclosure for that transaction. (Performance may be
+  improved if they do cache witness data for spent notes.)
+
+
+Specification
+=============
+
+Payment disclosure data structure
+---------------------------------
+
+A payment disclosure has the following fields:
+
+- `txid`: Transaction id for the given transaction `tx`.
+
+- `msg`: A message field, which could contain a challenge value from the party to whom
+  the payment disclosure is directed.
+
+- `sapling_outputs`: A vector of the Sapling Outputs that we are disclosing
+  [0..tx.shieldedOutputs.len()]
+
+  - `index`: An index into `tx.shieldedOutputs`.
+  - `ock`: The outgoing cipher key that allows this output to be recovered.
+
+- `sapling_spends`: A vector of the Sapling Spends for which we are proving spend authority
+  [0..tx.shieldedSpends.len()]
+
+  - `index`: An index into `tx.shieldedSpends`.
+  - `cv`
+  - `rk`
+  - `zkproof_spend`
+  - [Optional] A payment address proof:
+    - Any `(d, pk_d)` such that `pk_d = [ivk] GH(d)`
+    - `nullifier_addr`
+    - `zkproof_addr`
+  - `spendAuthSig`
+
+- `transparent_inputs`: A vector of the transparent inputs for which we are proving spend
+  authority [0..tx.vin.len()]
+
+  - `index`: An index into `tx.vin`.
+  - `sig`: A BIP 322 signature.
+
+    - TODO: `zcashd` currently only supports the legacy format defined in BIP 322. We may
+      want to backport full BIP 322 support before having transparent input support in
+      this ZIP, to ensure it does what we need.
+    - TODO: BIP 322 specifies consensus rule checks as part of the signature verification
+      process. We will likely need to migrate it over to an equivalent ZIP that specifies
+      these for Zcash (which has a different set of script validation consensus rules).
```

Comment:
Yes, BIP 322 says "check the signature using consensus rules, then upgradable rules", but Zcash AFAIK has no such thing as upgradeable rules.